internal/generate/olm-catalog/descriptor/descriptor.go

```diff
@@ -51,19 +51,50 @@ func GetCRDDescriptionForGVK(apisDir string, gvk schema.GroupVersionKind) (olmap
 	if strings.Contains(group, ".") {
 		group = strings.Split(group, ".")[0]
 	}
-	apiDir := filepath.Join(apisDir, group, gvk.Version)
-	universe, err := getTypesFromDir(apiDir)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return olmapiv1alpha1.CRDDescription{}, ErrAPIDirNotExist
-		}
-		return olmapiv1alpha1.CRDDescription{}, err
+
+	// Check if apisDir exists
+	if exists, _ := isDirExist(apisDir); !exists {
+		return olmapiv1alpha1.CRDDescription{}, ErrAPIDirNotExist
 	}
-	apiPkg := path.Join(projutil.GetGoPkg(), filepath.ToSlash(apiDir))
-	pkgTypes, err := getTypesForPkg(apiPkg, universe)
+
+	// Check if the kind pkg is at the expected layout
+	// multi-group layout: <api-dir>/<group>/<version>
+	// single-group layout: <api-dir>/<version>
+	expectedPkgPath, err := getExpectedPkgLayout(apisDir, group, gvk.Version)
 	if err != nil {
 		return olmapiv1alpha1.CRDDescription{}, err
 	}
+
+	// Get pkg types for the given GVK
+	var pkgTypes []*types.Type
+	if expectedPkgPath != "" {
+		// Look for the pkg types at the expected single or multi group import path
+		universe, err := getTypesFromDirRecursive(expectedPkgPath)
+		if err != nil {
+			return olmapiv1alpha1.CRDDescription{}, err
+		}
+		pkgTypes, err = getTypesForPkgPath(expectedPkgPath, universe)
+		if err != nil {
+			return olmapiv1alpha1.CRDDescription{}, err
+		}
+	} else {
+		// Unknown apis directory layout: <apis-dir>/.../<version>
+		// Look in <api-dir> recursively for expected pkg name <version>
+
+		// TODO: gengo.parse.AddDirRecursive() will (sometimes?) fail if the
+		// root apisDir has no .go files.
+		// Workaround for this is to have a doc.go file in the package.
+		// Move away from using gengo in the future if possible.
```

Comment:
Yes I'd very much like to move to controller-tools' parser implementation instead.